python/mxnet/ndarray/ndarray.py

```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
+            if indexing_dispatch_code == _BASIC_INDEXING_STEP_EQUAL_ONE:
+                return self._basic_indexing_step_equal_one(key)
+            elif indexing_dispatch_code == _BASIC_INDEXING_STEP_NOT_ONE:
+                return self._basic_indexing_step_not_one(key)
+            elif indexing_dispatch_code == _ADVANCED_INDEXING:
+                return self._advanced_indexing(key)
+            else:
+                raise ValueError('Indexing NDArray with index=%s is not supported' % str(key))
+        else:
+            raise ValueError('NDArray does not support slicing with key %s of type %s'
+                             % (str(key), str(type(key))))
+
+    def _advanced_indexing(self, key):
+        """Get item when key is a tuple of any objects of the following types:
+        NDArray, np.ndarray, list, tuple, slice, and integer. Note that if
+        all the objects in the tuple are of type slice with step=1, it should
+        invoke _basic_indexing_step_equal_one for better efficiency."""
+        def _is_advanced_index(index):
+            """The definition of advanced index here includes integers as well, while
+            integers are considered as basic index type when the key contains only
+            slices and integers."""
+            return not isinstance(index, py_slice)
+
+        if not isinstance(key, tuple):
+            raise ValueError('index=%s must be a tuple type to use advanced indexing,'
+                             ' received type=%s' % (str(key), str(type(key))))
+        assert len(key) > 0, "Cannot slice with empty indices"
+        shape = self.shape
+        assert len(shape) >= len(key),\
+            "Slicing dimensions exceeds array dimensions, %d vs %d" % (len(key), len(shape))
+        indices = []
+        dtype = 'int32'
+        need_broadcast = (len(key) != 1)
+        advanced_indices = []  # include list, NDArray, np.ndarray, integer
+        basic_indices = []  # include only slices
+        advanced_index_bshape = None  # final advanced index shape
+        for i, idx_i in enumerate(key):
+            is_advanced_index = True
+            if isinstance(idx_i, (np.ndarray, list, tuple)):
+                idx_i = array(idx_i, ctx=self.context, dtype=dtype)
+                advanced_indices.append(i)
+            elif isinstance(idx_i, py_slice):
+                start, stop, step = _get_index_range(idx_i.start, idx_i.stop, shape[i], idx_i.step)
+                idx_i = arange(start, stop, step, ctx=self.context, dtype=dtype)
+                basic_indices.append(i)
+                is_advanced_index = False
+            elif isinstance(idx_i, integer_types):
+                start, stop, step = _get_index_range(idx_i, idx_i+1, shape[i], 1)
+                idx_i = arange(start, stop, step, ctx=self.context, dtype=dtype)
+                advanced_indices.append(i)
+            elif isinstance(idx_i, NDArray):
+                dtype = idx_i.dtype
```

Comment:
different indices could have inconsistent type.

Reply:
I forced all dtypes to be 'int32' and added a unit test case.